tools/src/wyvern/tools/parsing/coreparser/WyvernParser.jj

```diff
@@ -378,6 +380,29 @@ AST TypeDeclaration() :
   }
 }
 
+AST DatatypeDeclaration() :
+{ Token name; AST body = null; AST metadata = null; Type extendedType=null; List<Type> comprisesList=null; boolean isTagged = false;
+  Token selfName = null;
+  Object tagInfo = null; boolean isResource = false;
+}
+{
+  [<TAGGED> { isTagged = true; } ]
+  [<RESOURCE> { isResource = true; } ]
+  <DATATYPE> name=<IDENTIFIER>
+  [ <EXTENDS> extendedType=Type() ]
```

Comment:
Also take `EXTENDS` and `COMPRISES` clauses out of the parser.  But treat the datatype as if it comprises all of the constructors.